src/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjection.java

```diff
@@ -0,0 +1,494 @@
+			{"_id);}, function(inj) { sleep("+INSERT_TIME_SHORT+");return 1;}, { out: 'x'}); "+
+			"db.injection.mapReduce(function() { emit(1,1",	
+			"_id);}, function(inj) { sleep("+INSERT_TIME_LONG+"); return 1;}, { out: 'x'}); "+
+			"db.injection.mapReduce(function() { emit(1,1"}};
+	private static final String[][] JSON_INJECTION = {{"$ne", "0"}, {"$gt", ""}, {"$regex", ".*"}};
+	// Error messages that addressing to a well-known vulnerability
+	private  final Pattern[] errorPatterns = {
+			Pattern.compile("RuntimeException: SyntaxError: unterminated string literal", Pattern.CASE_INSENSITIVE),
+			Pattern.compile("MongoResultException", Pattern.CASE_INSENSITIVE)};
+	// Log prints
+	private static final String JSON_EX_LOG = "try to convert the payload in json format";
+	private static final String IO_EX_LOG = "try to send an http message";
+	private static final String URI_EX_LOG = "try to get the message's Uri";
+	private static final Logger LOG = Logger.getLogger(MongoDbInjection.class);
+
+	@Override
+	public int getCweId() {
+		return 943;
+	}
+
+	@Override
+	public int getWascId() {
+		return 19;
+	}
+
+	public int getId() {
+		return 40033;
+	}
+
+	public Tech getTech() {
+		//TODO change in Tech.MongoDB as soon as available
+		return Tech.Db;
+	}
+
+	@Override
+	public boolean targets(TechSet technologies) {
+		//TODO change in Tech.MongoDB as soon as available
+		return technologies.includes(Tech.Db);
+	}
+
+	@Override
+	public String getName() {
+		return Constant.messages.getString(MESSAGE_PREFIX + "name");
+	}
+
+	@Override
+	public String getDescription() {
+		return Constant.messages.getString(MESSAGE_PREFIX + "desc");
+	}
+
+	@Override
+	public int getCategory() {
+		return Category.INJECTION;
+	}
+
+	@Override
+	public String getSolution() {
+		return Constant.messages.getString(MESSAGE_PREFIX + "soln");
+	}
+
+	@Override
+	public String getReference() {
+		return Constant.messages.getString(MESSAGE_PREFIX + "refs");
+	}
+
+	@Override
+	public String[] getDependency() {
+		return new String[] {};
+	}
+
+	public String getExtraInfo(String attack) {
+		return Constant.messages.getString(MESSAGE_PREFIX + "extrainfo." + attack);
+	}
+
+	@Override
+	public void init() {
+		if (LOG.isDebugEnabled()) {
+			LOG.debug("Initialising MongoDB penertration tests");
+		}
+		if (this.getAttackStrength() == AttackStrength.LOW) {
+			SLEEP_TIME_SHORT = 300;
+			SLEEP_TIME_LONG = 600;
+			doCrashScan = false;
+			doAllDataScan = true;
+			doTimedScan = true;
+			doJsonScan = true;
+			getMoreConfidence = false;
+			doAuthBypass = true;
+		}
+		else if (this.getAttackStrength() == AttackStrength.MEDIUM) {
+			SLEEP_TIME_SHORT = 400;
+			SLEEP_TIME_LONG = 800;
+			doCrashScan = true;
+			doAllDataScan = true;
+			doTimedScan = true;
+			doJsonScan = true;
+			getMoreConfidence = true;
+			doAuthBypass = true;
+		}
+		else if (this.getAttackStrength() == AttackStrength.INSANE) {
+			SLEEP_TIME_SHORT = 500;
+			SLEEP_TIME_LONG = 1500;

```

Comment:
I'm worried about these (and above) sleep values. I believe Mongo uses milliseconds in it's sleep command. .5sec to 1.5sec differences _**might**_ be reliably detectable on localhost scenarios but I'm not sure they will be across a network or on a system that has multiple users.

Might need something like:
https://github.com/zaproxy/zap-extensions/pull/645
Based on:
https://github.com/zaproxy/zaproxy/pull/2648/files
https://github.com/zaproxy/zaproxy/pull/3079
https://github.com/zaproxy/zaproxy/pull/3081/files

Reply:
Oh, great!. An estimate based on the average propagation time is first made and then a delay is added. I can do it! maybe adding a different delay depending on the power of the attack and in each case of the order of seconds. 
What do you think about the double test (in the aforementioned attack)? 
Thanks.

Reply:
Just keep in mind that the round trip time needs to stay under the default connection timeout. Which is 20s if I recall correctly.
https://github.com/zaproxy/zap-core-help/wiki/HelpUiDialogsOptionsConnection#timeout-in-seconds
(Help doc should probably list the default <grin> ... another item for my todo list)